src/storage/btree_load.c

```diff
@@ -3878,3 +3875,385 @@ btree_get_perf_btree_page_type (THREAD_ENTRY * thread_p, PAGE_PTR page_ptr)
+  (pk_isid).oid_list->max_oid_cnt = pk_isid.oid_list->capacity;
+  (pk_isid).oid_list->next_list = NULL;
+  (pk_isid).copy_buf_len = DBVAL_BUFSIZE;
+  (pk_isid).check_not_vacuumed = true;
+
+  /* Lock the primary key class. */
+  ret = lock_object (thread_p, sort_args->fk_refcls_oid, oid_Root_class_oid, SIX_LOCK, LK_COND_LOCK);
+  if (ret != LK_GRANTED)
+    {
+      ret = er_errid ();
+      goto end;
+    }
+
+  /* Primary key index search prepare */
+  if (btree_prepare_bts (thread_p, &pk_bt_scan, sort_args->fk_refcls_pk_btid, &pk_isid, NULL, NULL, NULL,
+			 NULL, NULL, false, NULL) != NO_ERROR)
+    {
+      assert (er_errid () != NO_ERROR);
+      goto end;
+    }
+
+  if (sort_args->key_type->is_desc)
+    {
+      is_desc = true;
+    }
+
+  /* Get the corresponding leaf of the foreign key. */
+  if (!is_desc)
+    {
+      /* Get first leaf vpid. */
+      vpid = load_args->vpid_first_leaf;
+    }
+  else
+    {
+      /* Get the last leaf. Current leaf is the last one. */
+      vpid = load_args->leaf.vpid;
+    }
+
+  while (true)
+    {
+      ret = advance_to_next_slot_and_fix_page (thread_p, sort_args->btid, &vpid, &curr_fk_pageptr, &fk_slot_id,
+					       &fk_key, sort_args->key_type->is_desc, &fk_key_cnt, &fk_header);
+      if (ret != NO_ERROR)
+	{
+	  goto end;
+	}
+
+      if (curr_fk_pageptr == NULL)
+	{
+	  /* Search has ended. */
+	  break;
+	}
+
+      /* We got the value from the foreign key, now search through the primary key index. */
+      found_p = false;
+      /* Search through the primary key index. */
+      if (pk_leaf == NULL)
+	{
+	  /* No search has been initiated yet, we start from root. */
+	  pk_leaf =
+	    btree_locate_key (thread_p, &(pk_bt_scan.btid_int), &fk_key, &(pk_bt_scan.C_vpid), &slot_id, &found_p);
+
+	  if (found_p == 0)
+	    {
+	      /* Value was not found at all, it means the foreign key is invalid. */
+	      val_print = pr_valstring (&fk_key);
+	      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_FK_INVALID, 2, sort_args->fk_name,
+		      (val_print ? val_print : "unknown value"));
+	      ret = ER_FK_INVALID;
+	      goto end;
+	    }
+
+	  /* Value was found, proceed with the next value */
+	  slot_id++;
+	}
+      else
+	{
+	  /* Get the header of the current leaf. */
+	  pk_header = btree_get_node_header (thread_p, pk_leaf);
+
+	  /* Get the number of keys in current leaf.  */
+	  pk_key_cnt = btree_node_number_of_keys (thread_p, pk_leaf);
+
+	  /* We try to resume the search in the current leaf. */
+	  while (!found_p && slot_id <= pk_key_cnt)
+	    {
+	      /* Try to get the value from the next slot. */
+	      ret = get_value_from_leaf_slot (thread_p, &(pk_bt_scan.btid_int), pk_leaf, slot_id, &pk_key);
+
+	      if (ret != NO_ERROR)
+		{
+		  goto end;
+		}
+
+	      /* We need to compare the current value with the new value from the primary key. */
+	      ret = btree_compare_key (&pk_key, &fk_key, pk_bt_scan.btid_int.key_type, 1, 1, NULL);
```

Comment:
Declare "DB_VALUE_COMPARE_RESULT compare_ret;" at the beginning of function and here do:
```c
	      /* We need to compare the current value with the new value from the primary key. */
	      compare_ret = btree_compare_key (&pk_key, &fk_key, pk_bt_scan.btid_int.key_type, 1, 1, NULL);
              if (compare_ret == DB_EQ)
                {
                  /* found, stop searching in pk */
                  break;
                }
              else if (compare_ret == DB_LT)
                {
                  /* continue looking */
                  continue;
                }
              else
                {
                  /* set fk_invalid error */
                  /* goto end */
                }
```